runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java

```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
      * @param entry entry to potentially apply
      * @return if we need to apply the entry.
      */
-    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry) {
+    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry, boolean isCheckpointEntry) {
+        // If white list mode, ignore everything that is not in the list or checkpoints
+        if (whiteList && !streamsToLoad.contains(streamId)) {
```

Comment:
This is kind of superfluous, only the second clause is needed. Also, why negate/negate?

If (streamsToLoad.contains(streamId)) return true;

Reply:
The logic is a bit more complicated. The last condition should always be checked, whether we are in whitelist or blacklist. It checks that we are not processing an entry that is already contained in a checkpoint we already processed.

This is why if streamsToLoad.contains(streamId) return true is actually not enough. I had this at the beginning, and it was not passing the tests.

I will re-work this function to make it more self-explanatory. I reckon it is a bit hard to read.